### PR TITLE
Feature: `after_model_change` response

### LINKED ISSUE #1029: Allow `insert_model` / `after_model_change` to control the create response
### Checklist

- [x] There are no similar issues or pull requests for this yet.

###  Problem

When overriding `insert_model` in a `ModelView`, the create handler in `application.py` always issues a `302` redirect after a successful insert:

```python
obj = await model_view.insert_model(request, form_data_dict)
url = self.get_save_redirect_url(...)
return RedirectResponse(url=url, status_code=302)
```

There is no way for a `ModelView` to stay on the create page or render a custom response after a successful create. The `after_model_change` hook returns `None` and cannot influence the HTTP response.

The only current workaround is to raise an exception from `insert_model`, which abuses exception-based control flow and causes the exception to be logged via `logger.exception(e)` in the create handler's except block, which makes no sense at all.

###  Use Case

One-time secret display after creating a resource (similar to how GitHub, and other platforms handle token creation). This is a common pattern for any admin who manages secrets, tokens, or credentials:

1. Admin creates an API key
2. The generated secret is shown once on the create page with a copy button
3. After navigating away, the secret is never shown again

###  Related Issues

#971 is somehow related. However, flash messages are designed for transient notifications shown on the next page after a redirect. A one-time secret needs to:

- Stay on the current page (no redirect)
- Be displayed prominently and persistently (not auto-dismissing)
- Include a copy-to-clipboard button
- Only disappear when the user explicitly navigates away

### Describe the solution you would like.

Make the return value of `after_model_change` meaningful. Currently, it always returns `None`. Instead, allow it to return:

- `None` - default redirect (fully backward-compatible)
- `dict` - re-render the create/edit template with extra context merged in
- `Response` - return a custom Starlette Response directly

### Describe alternatives you considered

- Raising an exception from `insert_model` to prevent the redirect - works but abuses control flow and pollutes logs via `logger.exception(e)`
- Flash messages ( #971 ) - insufficient for one-time secret display since they redirect and auto-dismiss

### Additional context

_No response_

I will happily contribute to this and give it a try.

### PR DESCRIPTION
**Pull request by dhcsousa**
Closes #1029 

## Summary

Makes the return value of `after_model_change` meaningful, allowing it to control the HTTP response after create/edit operations.

### `after_model_change` new return types:
- **`None`** (default) — standard 302 redirect (fully backward-compatible)
- **`Response`** — returns a custom Starlette `Response` directly

### Status

- [x] Tests included
- [x] Docs updated (docstring, configurations, cookbook)
- [x] Python 3.9–3.14 compatible

### CODE REVIEW

**Comment by dhcsousa** [comment]
@mmzeynalli Here's the PR. Note: the docs/cookbook content was drafted with LLM assistance.

**Comment by dhcsousa** [comment]
@mmzeynalli, solved the PR conflicts with the new version, let me know what you think 😄 .

**Comment by mmzeynalli** [comment]
Thanks a lot. I did not review this because I was focusing on new version release. As that is done, I will continue with 
reviews. Thanks beforehand for your effort!

**Comment by dhcsousa** [comment]
> Thanks a lot. I did not review this because I was focusing on new version release. As that is done, I will continue with reviews. Thanks beforehand for your effort!

No worries, let me know if I can help in anything else!

**Comment by dhcsousa** [comment]
Hi @mmzeynalli, just checking in on this PR whenever you have time to review it. Do you think this feature could be ready for an upcoming release, or is there anything else I should adjust first? Thanks!

**Comment by mmzeynalli** [comment]
@dhcsousa Hi. Sorry, I have been out of country and have been busy. I know about the issues, and hopefully I will be back in 1-2 weeks to work on new version. If update needed, I will notify you.

**Comment by mmzeynalli** [comment]
@dhcsousa can this problem be solved with flash messages?

**Comment by dhcsousa** [comment]
> @dhcsousa can this problem be solved with flash messages?

Hi @mmzeynalli, thanks for taking a look. I don't think flash messages fully cover this; the main motivating case (the one-time secrets cookbook) is specifically about **not** persisting the value anywhere. 

If I am not mistaken, flash messages are stored in `request.session["_messages"]`, which defeats the point.

**Comment by mmzeynalli** [comment]
Okay, I will check that too

File: sqladmin/application.py [hunk 1]
```diff
@@ -614,6 +614,16 @@ async def create(self, request: Request) -> Response:
...
 
+        after_response = getattr(request.state, "_sqladmin_after_change_response", None)
+        if isinstance(after_response, Response):
+            return after_response
+        if isinstance(after_response, dict):
+            context["obj"] = obj
+            context.update(after_response)
+            return await self.templates.TemplateResponse(
+                request, model_view.create_template, context
+            )
```
Comment (mmzeynalli) [inline sqladmin/application.py]: Lets drop dict support. This won't be used a lot, so user needing custimzation can achieve the same output with Response

File: sqladmin/application.py [hunk 2]
```diff
@@ -670,6 +680,16 @@ async def edit(self, request: Request) -> Response:
...
 
+        after_response = getattr(request.state, "_sqladmin_after_change_response", None)
+        if isinstance(after_response, Response):
+            return after_response
+        if isinstance(after_response, dict):
+            context["obj"] = obj
+            context.update(after_response)
+            return await self.templates.TemplateResponse(
+                request, model_view.edit_template, context
+            )
```
Comment (mmzeynalli) [inline sqladmin/application.py]: same

File: sqladmin/widgets.py [hunk 3]
```diff
@@ -119,6 +119,8 @@ def __call__(self, field: Field, **kwargs: Any) -> Markup:
...
             kwargs["checked"] = True
 
-        return Markup(
-            '<div class="form-switch d-flex align-items-center h-100">%s</div>'
-        ) % Markup.escape(super().__call__(field, **kwargs))
+        return Markup(  # nosec B704
+            '<div class="form-switch d-flex align-items-center h-100">'
+            + str(Markup.escape(super().__call__(field, **kwargs)))
+            + "</div>"
+        )
```
Comment (mmzeynalli) [inline sqladmin/widgets.py:126]: This is not complaining, why add nosec?

**Review by mmzeynalli — CHANGES_REQUESTED** [review]
Let's change multiple things:

1. Drop dict support, only None or Response
2. Lets show secret as popup/modal or something else. Have it in layout.html, with `if secret` condition. One might want to use it in BaseView as well.
3. It would be nice to include screenshots how it is seen

File: sqladmin/widgets.py (hunk 3 above)
Comment (dhcsousa) [reply sqladmin/widgets.py:126]: Well, actually, I get the `bandit` issue:

```
format...................................................................Passed
lint.....................................................................Passed
secure...................................................................Failed
- hook id: secure
- exit code: 2

uv run bandit -r sqladmin --config pyproject.toml
[main]  INFO    profile include tests: None
[main]  INFO    profile exclude tests: None
[main]  INFO    cli include tests: None
[main]  INFO    cli exclude tests: None
[main]  INFO    using config: pyproject.toml
[main]  INFO    running on Python 3.13.3
[tester]        WARNING nosec encountered (B704), but no failed test on file sqladmin/widgets.py:69
Run started:2026-05-26 17:15:07.439841+00:00

Test results:
>> Issue: [B704:markupsafe_markup_xss] Potential XSS with ``markupsafe.Markup`` detected. Do not use ``Markup`` on untrusted data.
   Severity: Medium   Confidence: High
   CWE: CWE-79 (https://cwe.mitre.org/data/definitions/79.html)
   More Info: https://bandit.readthedocs.io/en/1.9.4/plugins/b704_markupsafe_markup_xss.html
   Location: sqladmin/widgets.py:122:15
121
122             return Markup(
123                 '<div class="form-switch d-flex align-items-center h-100">'
124                 + str(Markup.escape(super().__call__(field, **kwargs)))
125                 + "</div>"
126             )

--------------------------------------------------

Code scanned:
        Total lines of code: 4353
        Total lines skipped (#nosec): 0

Run metrics:
        Total issues (by severity):
                Undefined: 0
                Low: 0
                Medium: 1
                High: 0
        Total issues (by confidence):
                Undefined: 0
                Low: 0
                Medium: 0
                High: 1
Files skipped (0):
make: *** [secure] Error 1

docs.....................................................................Passed
test.....................................................................Passed
```

Comment (dhcsousa) [reply sqladmin/widgets.py:126]: I will leave the `nosec`, let me know if you really want me to remove it. Since it's using the `Markup` I don't see the issue.

File: sqladmin/application.py (hunk 1 above)
Comment (dhcsousa) [reply sqladmin/application.py]: Well, I think the `dict` would be useful as a one-line shortcut for adding extra context to the existing create/edit template without building a full `TemplateResponse`. Either way, I removed it, it's done :)

File: sqladmin/application.py (hunk 2 above)
Comment (dhcsousa) [reply sqladmin/application.py]: Done, please check.

**Comment by dhcsousa** [comment]
> Let's change multiple things:
> 
> 1. Drop dict support, only None or Response
> 2. Lets show secret as popup/modal or something else. Have it in layout.html, with `if secret` condition. One might want to use it in BaseView as well.
> 3. It would be nice to include screenshots how it is seen

Everything is done. For 3 here you go:

one-time-secret-demo

File: sqladmin/templates/sqladmin/secret_modal.html [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    var modalEl = document.getElementById('modal-secret');
+    document.getElementById('modal-secret-trigger').click();
+    var nextUrl = modalEl.dataset.nextUrl;
+    if (nextUrl) {
+      modalEl.addEventListener('hidden.bs.modal', function () {
+        window.location.replace(nextUrl);
+      });
+    }
+  });
+</script>
```
Comment (mmzeynalli) [inline sqladmin/templates/sqladmin/secret_modal.html]: Better add this to js file

File: sqladmin/application.py [hunk 5]
```diff
@@ -670,6 +684,19 @@ async def edit(self, request: Request) -> Response:
...
+            return after_response
+
+        if getattr(request.state, "sqladmin_secret", None):
+            context["obj"] = obj
+            context["secret_next_url"] = str(
+                request.url_for("admin:list", identity=identity)
+            )
+            return await self.templates.TemplateResponse(
+                request, model_view.edit_template, context
+            )
```
Comment (mmzeynalli) [inline sqladmin/application.py]: repeated, as above. Create helper function for this

File: sqladmin/flash.py [hunk 6]
```diff
@@ -114,6 +114,28 @@ def success(cls, request: Request, message: str, title: str = "") -> bool:
...
+        does not survive a redirect. Callers that want the modal to show on a
+        create/edit submission should call this from ``after_model_change``;
+        the create/edit handler will re-render instead of redirecting when a
+        secret is set.
+        """
+        request.state.sqladmin_secret = {
+            "value": value,
+            "title": title,
+            "label": label,
+        }
```
Comment (mmzeynalli) [inline sqladmin/flash.py]: Im not sure if I like adding secret to Flash, as it defeats the purpose of it. I think it would be better to create new class, and expect it to be extendable. However, I would still like to get @aminalaee 's opinion

**Review by mmzeynalli — CHANGES_REQUESTED** [review]
Let's also add these, for extra security

```
    @staticmethod
    def _apply_no_store_headers(response: Response) -> None:
        response.headers["Cache-Control"] = (
            "no-store, no-cache, must-revalidate, max-age=0"
        )
        response.headers["Pragma"] = "no-cache"
        response.headers["Expires"] = "0"
```

File: sqladmin/flash.py (hunk 6 above)
Comment (dhcsousa) [reply sqladmin/flash.py]: I agrre. `Flash` is session-backed if I'm not mistaken. I will make a new `Reveal` that stashes one-time, request-related values on `request.state` (never the session) for rendering on the next response? 

Then I will start with `Reveal.secret(...)`, which can be expanded later if needed.

Comment (mmzeynalli) [reply sqladmin/flash.py]: Maybe, name it Secret.reveal_once?

File: sqladmin/application.py (hunk 5 above)
Comment (dhcsousa) [reply sqladmin/application.py]: Done, please check 😄 .

**Comment by dhcsousa** [comment]
> Let's also add these, for extra security
> 
> ```
>     @staticmethod
>     def _apply_no_store_headers(response: Response) -> None:
>         response.headers["Cache-Control"] = (
>             "no-store, no-cache, must-revalidate, max-age=0"
>         )
>         response.headers["Pragma"] = "no-cache"
>         response.headers["Expires"] = "0"
> ```

Done, please check 😄 .

File: sqladmin/templates/sqladmin/secret_modal.html (hunk 4 above)
Comment (dhcsousa) [reply sqladmin/templates/sqladmin/secret_modal.html]: Done, please check 😄 .

File: sqladmin/flash.py (hunk 6 above)
Comment (dhcsousa) [reply sqladmin/flash.py]: Done, let me know what you guys think @mmzeynalli @aminalaee

File: sqladmin/application.py [hunk 7]
```diff
@@ -808,6 +848,14 @@ def _normalize_wtform_data(obj: Any) -> dict:
                 form_data[field_name + "_"] = value
         return form_data
 
+    @staticmethod
+    def _apply_no_store_headers(response: Response) -> None:
+        response.headers["Cache-Control"] = (
+            "no-store, no-cache, must-revalidate, max-age=0"
+        )
+        response.headers["Pragma"] = "no-cache"
+        response.headers["Expires"] = "0"
```
Comment (mmzeynalli) [inline sqladmin/application.py]: maybe have this at Secret class?

**Review by mmzeynalli — CHANGES_REQUESTED** [review]
Nicely done! One little pesky comment tho)

**Comment by mmzeynalli** [comment]
Add something like this to the docs as well:

```

When revealing secrets from a `BaseView`, apply anti-cache headers explicitly:

```python
response = await self.templates.TemplateResponse(request, "your_template.html", context)
Secret.apply_no_store_headers(response)
return response
```
```

Also, if possible do this is JS:

```js
  var copyButton = document.getElementById('modal-secret-copy');
  var secretValueInput = document.getElementById('modal-secret-value');
  if (copyButton && secretValueInput) {
    copyButton.addEventListener('click', function () {
      if (navigator.clipboard && navigator.clipboard.writeText) {
        navigator.clipboard.writeText(secretValueInput.value);
      }
    });
  }
  ```

And change button to:

`<button class="btn btn-outline-secondary" type="button" id="modal-secret-copy">`

**Comment by dhcsousa** [comment]
> Add something like this to the docs as well:
> 
> ```
> 
> When revealing secrets from a `BaseView`, apply anti-cache headers explicitly:
> 
> ```python
> response = await self.templates.TemplateResponse(request, "your_template.html", context)
> Secret.apply_no_store_headers(response)
> return response
> ```
> 
> ```
> 
> Also, if possible do this is JS:
> 
> ```js
>   var copyButton = document.getElementById('modal-secret-copy');
>   var secretValueInput = document.getElementById('modal-secret-value');
>   if (copyButton && secretValueInput) {
>     copyButton.addEventListener('click', function () {
>       if (navigator.clipboard && navigator.clipboard.writeText) {
>         navigator.clipboard.writeText(secretValueInput.value);
>       }
>     });
>   }
> ```
> 
> And change button to:
> 
> `<button class="btn btn-outline-secondary" type="button" id="modal-secret-copy">`

It's done 😄

File: sqladmin/application.py (hunk 7 above)
Comment (dhcsousa) [reply sqladmin/application.py]: It's done 😄

**Comment by dhcsousa** [comment]
@mmzeynalli, I have implemented all that was suggested. Please review the open threads and close all that are already resolved 😄.

**Review by mmzeynalli — APPROVED** [review]
Nice! Thanks a lot!
